### ISSUE TITLE
[HistogramBarSeries] Does not render bar charts when `g` prop is an empty string

### ISSUE THREAD

**Issue by janmonschke** [elastic/elastic-charts#1480]
**Describe the bug**
When passing data with an empty `g` props to the `HistogramBarSeries` component it only renders the `No data to display` message.

**To Reproduce**
I created a test case on codesandbox here: https://codesandbox.io/s/cocky-tesla-9vlu3?file=/src/App.tsx

The first chart has the data where `g` is empty, the second one has `g` set to `'empty'`.

**Expected behaviour**
I am not sure if this is actually a bug or not. The fact that the component renders the `No data to display` message makes me think that it actually is because there is data to display. It's just missing a label to display.

**Screenshots**
<img width="1459" alt="Screenshot 2021-11-17 at 11 58 35" src="https://user-images.githubusercontent.com/68591/142188148-5eb17a49-3993-4148-976a-91727183ae09.png">

<img width="250" alt="Screenshot 2021-11-17 at 12 00 55" src="https://user-images.githubusercontent.com/68591/142188545-a3bb0b5f-4044-4275-a024-6fc3fcccccae.png">

**Kibana Cross Issues**
https://github.com/elastic/kibana/issues/111670
https://github.com/elastic/kibana/pull/118868

**Comment by nickofthyme** [comment]
Yeah I think this is intended behavior. We could make it to where it displays the string `'(empty)'` like in Lens but then we need to add another prop in case someone wants another string. We also do something similar in `vis_type_xy` when no x metric is set we default to `xAccessor={() => '_all'}`, see code [here](https://github.com/elastic/kibana/blob/c983bd94ddaccd69091a0663b85c0ecab88ca855/src/plugins/vis_types/xy/public/utils/accessors.tsx#L18-L23). I see an empty string as very similar to `undefined` or `null` is most cases, so I believe this behavior is a suitable default.

That being said, you can explicitly define a string for empty cases by passing an `AssessorFn`, works for any accessor of an xy chart series, to check the value and conditionally return anything. Something like...

```ts
const splitSeriesAccessors = React.useMemo<AccessorFn[]>(
  () => [(datum: any) => (datum.g === "" ? datum.g : "(empty)")],
  []
);
```

The callback argument is each datum/row in the provided data array. See in action [here](https://codesandbox.io/s/agitated-microservice-ej8n4?file=/src/App.tsx). Let me know if that helps.

**Comment by janmonschke** [comment]
Thanks for the explanation @nickofthyme 👍 

The data is formatted before in the case of the histogram that I'm fixing so I added the logic to provide the empty label to that fn: https://github.com/elastic/kibana/pull/118868 .

I think in the long run it might make sense to move that logic to an `AccessorFn` as you proposed. That would require some refactoring of our custom logic for the chart legends. I will add that refactoring to our next batch of work :)

**Comment by nickofthyme** [comment]
Yeah using functions instead of string keys for any accessor can add some refactoring work.

For example in `vis_type/xy` we have split series that can be range aggs which return `{ start: number, end: number }` which needs to be `'${start} to ${end}'`. So we use `AssessorFn` there to do `(d) => '${d.start} to ${d.end}'` or something similar.

The issue is anytime you need to know the split accessor of a given series, say clicking a bar, you need to determine what `AccessorFn` it belongs to. In charts we process the data given the provided accessors so we don't know the data groupings ahead of time, this is why we pass a `SeriesIdentifier` in the callback for interaction event, for the user to check "is this using this x/y/split/etc. accessor? then do this".

Split accessors are a bit different than x and y because they determine the grouping not the value. As such, it requires some clever logic to find which split accessor the interaction belongs to. You can look at the `transform_click_event.ts` logic in the charts plugin [here](https://github.com/elastic/kibana/blob/c983bd94ddaccd69091a0663b85c0ecab88ca855/src/plugins/charts/public/static/utils/transform_click_event.ts#L46-L68) to get an idea. My hope is that this code can be shared to alleviate this issue for most chart users.

It's hard to explain so forgive me if that confusing, happy to jump on a call if/when you decide to go with the `AccessorFn` approach. 👍🏼

**Comment by markov00** [comment]
@nickofthyme I think we can however find and fix this behavior, or at least correct part of it.
In particular the `isChartEmpty` selector is bound to the legend items. The dataset expressed as in the example is correctly computed and the geometries are generated. 

What blocks the rendering is the `isChartEmpty` because the series name is now configured as an empty string and it discarded from the legend list:
https://github.com/elastic/elastic-charts/blob/962e089fe61540dcd5ed3612d41d915c29b98d13/packages/charts/src/chart_types/xy_chart/legend/legend.ts#L130

Removing the empty check makes the chart appear.
One more issue is on the tooltip, we have a similar control on the tooltip, where a tooltip value is shown only if it has a non-empty string:
https://github.com/elastic/elastic-charts/blob/962e089fe61540dcd5ed3612d41d915c29b98d13/packages/charts/src/chart_types/xy_chart/tooltip/tooltip.ts#L74

removing also that fixes the rendering
<img width="691" alt="Screenshot 2021-11-18 at 12 12 39" src="https://user-images.githubusercontent.com/1421091/142405538-7bd8edfb-4351-4f7d-9ca0-9d3ef6e2678c.png">

Is anyway not perfect, but at least doesn't say something wrong about the data.

I think we should also detach the `isChartEmpty` selector from the visible  legend list items and connect it to the list of detected DataSeries
